RecoEgamma/EgammaTools/src/EnergyScaleCorrection.cc

```diff
@@ -0,0 +1,431 @@
+  }  
+  //validate the result, just to be sure
+  if(!result.first->first.inCategory(runnr,et,eta,r9,gainSeed)){
+    throw cms::Exception("LogicError") <<" error found scale category "<<result.first->first<<" that does not contain run "<<runnr<<" et "<<et<<" eta "<<eta<<" r9 "<<r9<<" gain seed "<<gainSeed;
+  }
+  return &result.first->second;
+}
+
+const EnergyScaleCorrection::SmearCorrection* 
+EnergyScaleCorrection::getSmearCorr(unsigned int runnr, double et, double eta, double r9,
+				    unsigned int gainSeed) const
+{
+
+  // buld the category based on the values of the object
+  CorrectionCategory category(runnr, et, eta, r9, gainSeed);
+  auto result = std::equal_range(smearings_.begin(),smearings_.end(),category,Sorter<CorrectionCategory,SmearCorrection>()); 
+  auto nrFound = std::distance(result.first,result.second);
+  if(nrFound==0){
+    LOGWARN("EnergyScaleCorrection") << "[ERROR] Smear category not found: " << category << " Returning uncorrected value.";
+    return nullptr;
+  }else if(nrFound>1){
+    std::ostringstream foundCats;
+    for(auto it = result.first;it!=result.second;++it){
+      foundCats<<"    "<<it->first<<std::endl;
+    }
+    throw cms::Exception("ConfigError") <<" error smear category "<<category<<" has "<<nrFound<<" entries "<<std::endl<<foundCats.str();
+  }
+  //validate the result, just to be sure
+  if(!result.first->first.inCategory(runnr,et,eta,r9,gainSeed)){
+    throw cms::Exception("LogicError") <<" error found smear category "<<result.first->first<<" that does not contain run "<<runnr<<" et "<<et<<" eta "<<eta<<" r9 "<<r9<<" gain seed "<<gainSeed;
+  }
+  return &result.first->second;
+}
+
+
+void EnergyScaleCorrection::addScale(const std::string& category, int runMin, int runMax,  
+				     double energyScale, double energyScaleErrStat, 
+				     double energyScaleErrSyst, double energyScaleErrGain)
+{
+  
+  CorrectionCategory cat(category,runMin,runMax); // build the category from the string
+  auto result = std::equal_range(scales_.begin(),scales_.end(),cat,Sorter<CorrectionCategory,ScaleCorrection>());
+  if(result.first!=result.second){
+    throw cms::Exception("ConfigError") << "[ERROR] Category already defined! "<<cat;
+  }
+  
+  ScaleCorrection corr(energyScale,energyScaleErrStat,energyScaleErrSyst,energyScaleErrGain);
+  scales_.push_back({cat,corr});
+  std::sort(scales_.begin(),scales_.end(),Sorter<CorrectionCategory,ScaleCorrection>()); 
+  
+}
+
+void EnergyScaleCorrection::addSmearing(const std::string& category,int runMin, int runMax,
+					double rho, double errRho, 
+					double phi, double errPhi,
+					double eMean, double errEMean)
+{
+  CorrectionCategory cat(category);
+  
+  auto res = std::equal_range(smearings_.begin(),smearings_.end(),cat,Sorter<CorrectionCategory,SmearCorrection>()); 
+
+  if(res.first!=res.second) {
+    throw cms::Exception("EnergyScaleCorrection") << "[ERROR] Smearing category already defined "<<cat;
```

Comment:
"[ERROR]" can be dropped probably, it can't really get worse than an exception

Reply:
done